scrapy/spiders/crawl.py

```diff
@@ -105,15 +115,27 @@ def _requests_to_follow(self, response):
 
     def _callback(self, response, **cb_kwargs):
         rule = self._rules[response.meta["rule"]]
-        return self._parse_response(
+        return self.parse_with_rules(
             response, rule.callback, {**rule.cb_kwargs, **cb_kwargs}, rule.follow
         )
 
     def _errback(self, failure):
         rule = self._rules[failure.request.meta["rule"]]
         return self._handle_failure(failure, rule.errback)
 
+    def parse_with_rules(self, response, callback, cb_kwargs, follow=True):
+        cb_kwargs["direct_call"] = False
+        return self._parse_response(response, callback, cb_kwargs, follow)
+
     async def _parse_response(self, response, callback, cb_kwargs, follow=True):
+        if cb_kwargs.get("direct_call", True):
+            warnings.warn(
+                "The CrawlSpider._parse_response method is deprecated: it "
+                "will be removed in future Scrapy releases. Please use the "
+                "CrawlSpider.parse_with_rules method instead."
+            )
+        cb_kwargs.pop("direct_call", None)
```

Comment:
Interesting strategy. I think it might be OK to go this route, as long as we use a more descriptive and less public name, e.g. `_direct_parse_response_call`.

However, there may be better approaches here.

For example, you could define a new private method, move the body of `_parse_response` to that method, and call that method from both `_parse_method` and `parse_with_rules`. In `_parse_method`, however, you also log a warning. Then you do not need to mess up with cb_kwargs, which is meant to end users, not for internal handlings.

It might also be good to log these warnings only once, and not once per callback call. So, you could run the method inheritance check on `__init__`, and define a private boolean attribute in `__init__` as well to keep track of whether or not `_parse_response` has been called, to only log a warning on the first of such calls.

Finally, I think the `parse_with_rules` should be async as well. If you tried that already and had issues, mind that I think you need to await your internal call to `_parse_response` (or to the new private method) for things to work as expected. That is, rather than:

```python
    async def parse_with_rules(…):
        return self._parse_response(…)
```

I think you need:

```python
    async def parse_with_rules(…):
        return await self._parse_response(…)
```

Reply:
I took the approach that you suggested  but I was not able to make `parse_with_rules` and `_parse_response` async. It raises this error:
```
return await self._parse_with_rules(response, callback, cb_kwargs, follow)
TypeError: object async_generator can't be used in 'await' expression
```
if I put `async/await` in the methods. I think it is because of yield on the body of the original function. It is working with only the `_parse_with_rules` being async but I don't know how much the others functions not being async would affect the operation of other parts of the code. 

I also created 3 tests to check if the warnings would be raising properly. Should I add more to test other stuff?

Reply:
`_parse_with_rules` is an async generator so you cannot await it, you need:

```python
def _parse_response(self, response, callback, cb_kwargs, follow=True):
    ...
    async for request_or_item in self._parse_with_rules(response, callback, cb_kwargs, follow):
        yield request_or_item
```